In [40]:
import pandas as pd

# Load dataset
df = pd.read_csv("StudentsPerformance.csv")
df_copy=df

#Show first few rows
df.head(10)

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   
5  female        group B          associate's degree      standard   
6  female        group B                some college      standard   
7    male        group B                some college  free/reduced   
8    male        group D                 high school  free/reduced   
9  female        group B                 high school  free/reduced   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75  
5                    none          71             83             78  
6               completed          88             95             92  
7                    none          40             43             39  
8               completed          64             64             67  
9                    none          38             60             50

In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


In [53]:
df.shape

(1000, 8)

In [56]:
df['lunch'].unique()

array(['standard', 'free/reduced'], dtype=object)

In [6]:
df['race/ethnicity'].unique()

array(['group B', 'group C', 'group A', 'group D', 'group E'],
      dtype=object)

In [7]:
df['parental level of education'].unique()

array(["bachelor's degree", 'some college', "master's degree",
       "associate's degree", 'high school', 'some high school'],
      dtype=object)

In [8]:
# Show data types and non-null counts
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


In [10]:
# Check for null values
df.isnull().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

In [11]:
# Check for duplicate rows
duplicate_count = df.duplicated().sum()
print("Duplicate rows:", duplicate_count)

Duplicate rows: 0


In [63]:
# Simplify education levels
df['parental level of education'] = df['parental level of education'].replace({
    "some high school": "High School",
    "high school": "High School",
    "some college": "College",
    "associate's degree": "College",
    "bachelor's degree": "University",
    "master's degree": "University"
})

# Rename ethnic groups
df['race/ethnicity'] = df['race/ethnicity'].replace({
    "group A": "Group 1",
    "group B": "Group 2",
    "group C": "Group 3",
    "group D": "Group 4",
    "group E": "Group 5"
})

# Preview changes
df[['race/ethnicity', 'parental level of education']].head()

race/ethnicity parental level of education
0        Group 2                  University
1        Group 3                     College
2        Group 2                  University
3        Group 1                     College
4        Group 3                     College

In [13]:
# Calculate average score
df['average score'] = df[['math score', 'reading score', 'writing score']].mean(axis=1)
df[['math score', 'reading score', 'writing score', 'average score']].head()

math score  reading score  writing score  average score
0          72             72             74      72.666667
1          69             90             88      82.333333
2          90             95             93      92.666667
3          47             57             44      49.333333
4          76             78             75      76.333333

In [16]:
# Define performance category
def categorize(avg):
    if avg >= 80:
        return 'High'
    elif avg >= 60 and avg < 80:
        return 'Medium'
    else:
        return 'Low'

df['performance category'] = df['average score'].apply(categorize)
df[['average score', 'performance category']].head(10)

average score performance category
0      72.666667               Medium
1      82.333333                 High
2      92.666667                 High
3      49.333333                  Low
4      76.333333               Medium
5      77.333333               Medium
6      91.666667                 High
7      40.666667                  Low
8      65.000000               Medium
9      49.333333                  Low

In [17]:
df['performance category'].value_counts()

performance category
Medium    517
Low       285
High      198
Name: count, dtype: int64

In [64]:
df['parental level of education'].value_counts()

parental level of education
College        448
High School    375
University     177
Name: count, dtype: int64

In [18]:
# Compare scores based on test preparation
prep_effectiveness = df.groupby('test preparation course')['average score'].mean()
print("Preparation Effectiveness:\n", prep_effectiveness)

Preparation Effectiveness:
 test preparation course
completed    72.669460
none         65.038941
Name: average score, dtype: float64


In [65]:
# Preparation Effectiveness
df['prep_effective'] = df.apply(
    lambda row: 'Prepared' if row['test preparation course'] == 'completed' else 'Not Prepared',
    axis=1
)
df

gender race/ethnicity parental level of education         lunch  \
0    female        Group 2                  University      standard   
1    female        Group 3                     College      standard   
2    female        Group 2                  University      standard   
3      male        Group 1                     College  free/reduced   
4      male        Group 3                     College      standard   
..      ...            ...                         ...           ...   
995  female        Group 5                  University      standard   
996    male        Group 3                 High School  free/reduced   
997  female        Group 3                 High School  free/reduced   
998  female        Group 4                     College      standard   
999  female        Group 4                     College  free/reduced   

    test preparation course  math score  reading score  writing score  \
0                      none          72             72             74   
1                 completed          69             90             88   
2                      none          90             95             93   
3                      none          47             57             44   
4                      none          76             78             75   
..                      ...         ...            ...            ...   
995               completed          88             99             95   
996                    none          62             55             55   
997               completed          59             71             65   
998               completed          68             78             77   
999                    none          77             86             86   

    prep_effective  
0     Not Prepared  
1         Prepared  
2     Not Prepared  
3     Not Prepared  
4     Not Prepared  
..             ...  
995       Prepared  
996   Not Prepared  
997       Prepared  
998       Prepared  
999   Not Prepared  

[1000 rows x 9 columns]

## SQL Operations:

#### Import the cleaned dataset into a SQL database.

###### Write SQL queries to:

a) List top 5 students with highest average scores.

b) Find the average math, reading, and writing scores by gender.

c) Compare average scores of students who completed test prep vs. those who didn’t.

d) Count how many students fall into each performance category.

e) Identify which ethnic group has the highest average total score.

In [20]:
# Save cleaned file
df.to_csv("cleaned_students1.csv", index=False)

In [66]:
df_copy1=df
df_copy1

gender race/ethnicity parental level of education         lunch  \
0    female        Group 2                  University      standard   
1    female        Group 3                     College      standard   
2    female        Group 2                  University      standard   
3      male        Group 1                     College  free/reduced   
4      male        Group 3                     College      standard   
..      ...            ...                         ...           ...   
995  female        Group 5                  University      standard   
996    male        Group 3                 High School  free/reduced   
997  female        Group 3                 High School  free/reduced   
998  female        Group 4                     College      standard   
999  female        Group 4                     College  free/reduced   

    test preparation course  math score  reading score  writing score  \
0                      none          72             72             74   
1                 completed          69             90             88   
2                      none          90             95             93   
3                      none          47             57             44   
4                      none          76             78             75   
..                      ...         ...            ...            ...   
995               completed          88             99             95   
996                    none          62             55             55   
997               completed          59             71             65   
998               completed          68             78             77   
999                    none          77             86             86   

    prep_effective  
0     Not Prepared  
1         Prepared  
2     Not Prepared  
3     Not Prepared  
4     Not Prepared  
..             ...  
995       Prepared  
996   Not Prepared  
997       Prepared  
998       Prepared  
999   Not Prepared  

[1000 rows x 9 columns]

In [24]:
!pip install mysql-connector-python sqlalchemy

In [25]:
import mysql.connector
import sqlalchemy
print("All good!")

All good!


In [42]:
import pandas as pd
from sqlalchemy import create_engine

user = "root"
password = "Mrmvcyg%4025"  # ← Encode @ as %40
host = "localhost"
database = "student_db"

engine = create_engine(f"mysql+mysqlconnector://{user}:{password}@{host}/{database}")

print("done")

done


In [43]:
# Preview the table
df_sql = pd.read_sql("SELECT * FROM students LIMIT 10", engine)
df_sql.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        Group 2                  University      standard   
1  female        Group 3                     College      standard   
2  female        Group 2                  University      standard   
3    male        Group 1                     College  free/reduced   
4    male        Group 3                     College      standard   

  test_preparation_course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   average_score performance_category prep_effective  
0        72.6667               Medium   Not Prepared  
1        82.3333                 High       Prepared  
2        92.6667                 High   Not Prepared  
3        49.3333                  Low   Not Prepared  
4        76.3333               Medium   Not Prepared

In [45]:
#a) List top 5 students with highest average scores.
# Top 5 students
pd.read_sql("SELECT * FROM students ORDER BY average_score DESC LIMIT 5", engine)

gender race_ethnicity parental_level_of_education     lunch  \
0  female        Group 5                  University  standard   
1  female        Group 5                     College  standard   
2    male        Group 5                  University  standard   
3  female        Group 5                  University  standard   
4  female        Group 4                 High School  standard   

  test_preparation_course  math_score  reading_score  writing_score  \
0                    none         100            100            100   
1                    none         100            100            100   
2               completed         100            100            100   
3               completed          99            100            100   
4               completed          97            100            100   

   average_score performance_category prep_effective  
0       100.0000                 High   Not Prepared  
1       100.0000                 High   Not Prepared  
2       100.0000                 High       Prepared  
3        99.6667                 High       Prepared  
4        99.0000                 High       Prepared

## Key Observations

#### 1. All Top Students Have “Standard” Lunch
Indicates that better nutrition and economic stability may contribute to higher academic performance.

#### 2. High Representation from Group 5
4 out of 5 students are from Group 5 ethnicity.

Suggests this group might have better educational access, cultural emphasis on academics, or other advantages.

#### 3. Most Have Highly Educated Parents
3 of 5 students have parents with University-level education.

This supports the correlation between parental education and student success.

#### 4. Test Prep Not Always Essential
2 of the top 3 students did not complete test prep — yet scored 100 in all subjects.

Suggests that test prep boosts average performers, but top students may perform well without it.

#### 5. Balanced Gender Representation
4 females, 1 male → Females dominate this top-5 list.

May reflect broader female strength in reading and writing, or a specific trend in your dataset.

### Insights:
#### Factor	Implication:

##### Lunch Type             - Ensuring standard or nutritious meals may improve outcomes

##### Parental Education	    - Parent-focused literacy and awareness programs could uplift student scores

##### Ethnicity Group Trends	- Equity-focused policies should investigate underrepresented groups

##### Test Prep Programs	    - Still valuable, especially for mid/low performers

##### Gender Focus	        - Encourage subject-specific support based on gender gaps (if found overall)

In [46]:
#b) Find the average math, reading, and writing scores by gender
# Average scores by gender
pd.read_sql("""
    SELECT gender,
           AVG(math_score) AS avg_math,
           AVG(reading_score) AS avg_reading,
           AVG(writing_score) AS avg_writing
    FROM students
    GROUP BY gender
""", engine)

gender  avg_math  avg_reading  avg_writing
0  female   63.6332      72.6081      72.4672
1    male   68.7282      65.4730      63.3112

### Key Observations

#### 1. Males outperform females in math
Males average ~5.1 points higher than females in math.

Suggests a gender performance gap favoring males in quantitative subjects.

#### 2. Females outperform males in reading and writing
Females average ~7.1 points higher in reading and ~9.2 points higher in writing.

Indicates a stronger verbal/language performance among female students.

#### 3. Overall academic strengths vary by gender
Males: stronger in math

Females: stronger in reading & writing

This is a consistent pattern seen in many educational studies globally.

### Insights & Implications

#### Insight	-Gendered subject strength is evident	
Actionable Implication-Consider gender-specific teaching approaches or focus areas. 
E.g., encourage female confidence in math, and male literacy support.

#### Insight- Writing and reading gaps are significant	
Actionable Implication- Schools could implement reading/writing enhancement programs for boys.
#### Insight- Math gap is present but smaller	
Actionable Implication- Encourage math engagement for girls through clubs, mentorships, or gamified learning.


In [48]:
#c) Compare average scores of students who completed test prep vs. those who didn’t.
# Students with and without test prep
pd.read_sql("""
    SELECT test_preparation_course,
           AVG(math_score) AS avg_math,
           AVG(reading_score) AS avg_reading,
           AVG(writing_score) AS avg_writing
    FROM students
    GROUP BY test_preparation_course
""", engine)

test_preparation_course  avg_math  avg_reading  avg_writing
0                    none   64.0779      66.5343      64.5047
1               completed   69.6955      73.8939      74.4190

### Key Observations
#### 1. Students who completed test prep scored significantly higher
+5.6 points in math

+7.4 points in reading

+9.9 points in writing

#### 2. Largest improvement is in writing
Indicates that test preparation may focus heavily on writing skills (e.g., essay writing, grammar structure, etc.)

#### 3. Consistent gains across all subjects
Not just a one-subject benefit — shows a well-rounded improvement across quantitative and language domains.

### Insights & Educational Implications
##### Insight: 
##### Test prep boosts scores across all subjects	
 Recommendation- Schools should encourage or subsidize test prep programs, especially for low/mid performers.
##### Nearly 10-point improvement in writing
Recommendatio- Could highlight writing as a skill that benefits most from structured guidance. Prioritize writing-intensive prep.
##### Strong case for data-driven interventions	
Recommendation- Use this evidence to advocate test prep as a strategic investment in school improvement programs.


In [49]:
d) Count how many students fall into each performance category.
#Count how many students fall into each performance category.
pd.read_sql("""SELECT performance_category, count(performance_category) As student_count FROM students
GROUP BY performance_category""", engine)

performance_category  student_count
0               Medium            517
1                 High            198
2                  Low            285

### Key Observations
##### 1. Majority of students are in the Medium category
517 out of 1000 students (51.7%) fall into the Medium performance tier.

Indicates that most students are performing close to average, with potential to improve with minimal intervention.

##### 2. Low performers represent a sizable minority
285 students (~28.5%) are categorized as Low performers.

This is a significant proportion that might be at risk of academic failure or falling behind.

##### 3. High performers are the smallest group
Only 198 students (~19.8%) fall into the High category.

Suggests that high academic excellence is relatively rare and may depend on multiple supportive factors (e.g., test prep, family education, school quality).

### Insights & Educational Implications
Insight	Recommendation
##### Most students are in the Medium range	
###### Recommendation- 
Focus improvement efforts here—this group can move to High with targeted support (e.g., mentorship, study plans).
##### High achievers are few	
##### Recommendation- 
Identify and nurture high performers through enrichment programs, scholarships, etc.
###### Low performers are significant	
###### Recommendation- 
Deploy remedial programs (e.g., tutoring, test prep, parental engagement) to prevent long-term academic issues.
##### Category-based analysis is essential	
##### Recommendation-
This tiered view enables data-driven segmentation for school interventions.


In [50]:
#Identify which ethnic group has the highest average total score.
pd.read_sql("""
    SELECT race_ethnicity, AVG(average_score) FROM students
GROUP BY race_ethnicity
ORDER BY AVG(average_score) DESC LIMIT 1
""", engine)

race_ethnicity  AVG(average_score)
0        Group 5           72.752381